Allow the Input/EditableText widget to scroll horizontally

jason-simmons opened the pull request:
EditableText is now rendered using a custom RenderObject
(RenderEditableParagraph).  RenderEditableParagraph draws the cursor,
handles scroll offsets, and provides feedback about the size of the text for
use by the scroll behavior.

jason-simmons commented:
Some questions:

* Should RenderEditableParagraph be a subclass of RenderParagraph, or would it be cleaner to fork RenderParagraph?

* Does the adjustment for the scroll offset in RenderEditableParagraph.paint look correct?
I saw some subtleties around this in RenderViewport and wasn't sure if RenderEditableParagraph needed to do something similar.

abarth commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,106 @@
+
+  Color get cursorColor => _cursorColor;
+  void set cursorColor(Color cursorColor) {
+    _cursorColor = cursorColor;
```
In these setters, we usually check whether the old value is equal to the new value and skip the markNeedsPaint call in that case.  We do that because the widget layer generates unconditional writes into the render tree and we don't want to mark ourselves as needing paint when its not actually needed.

Hixie replied:
Use "value" or "newCursorColor" to avoid shadowing "cursorColor" here. Generally I prefer "value" because it makes copy/pasting setters easier later.

abarth commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,106 @@
+
+  bool get showCursor => _showCursor;
+  void set showCursor(bool showCursor) {
+    _showCursor = showCursor;
```
ditto

abarth commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,106 @@
+    markNeedsPaint();
+  }
+
+  bool allowLineWrap() => false;
```
`bool get allowLineWrap => false;`

That way it's a property (more idomatic)

abarth commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,106 @@
+    layoutText(constraints);
+
+    context.canvas.save();
+    context.canvas.clipRect(offset & size);
```
We should skip the clip rect if the _contentSize.width <= size.width.  The clip generates costs layer in the pipeline.

abarth commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
 // values up to the nearest whole pixel value. The right long-term fix is to do
 // layout using fixed precision arithmetic.
-double _applyFloatingPointHack(double layoutValue) {
+double applyFloatingPointHack(double layoutValue) {
```
Oh, it's sad to make this public.  Maybe we should push this logic down into TextPainter.

Hixie replied:
Either push the logic that uses applyFloatingPointHack into TextPainter as per adam's comment, or file a bug and mention it in the checkin comment.

abarth commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
 
-  TextPainter _textPainter;
+  // Exposed for use by subclasses.
+  TextPainter textPainter;
```
final ?

abarth commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
   }
 
-  TextPainter _textPainter;
+  // Exposed for use by subclasses.
```
I'd skip comments like these given that there's no way to enforce them.

abarth commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
   }
 
-  void _layout(BoxConstraints constraints) {
+  // Can be overridden by subclasses.
```
I'd skip this comment too.  Everything can be overridden by subclasses.

abarth commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
 
-  void _layout(BoxConstraints constraints) {
+  // Can be overridden by subclasses.
+  bool allowLineWrap() => true;
```
Why not take this value as a constructor parameter and provide a getter and a setter?  As written, the value is neither final nor is there a notification pathway for when it changes.  We should either make it final (e.g., by taking the value as a constructor parameter and not providing a setter) or have a notification pathway for it changing (e.g., by having a setting that calls markNeedsLayout as appropriate).

abarth commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
+  // Can be overridden by subclasses.
+  bool allowLineWrap() => true;
+
+  // Exposed for use by subclasses.
```
I'd skip these comments.

Rather than exposing all these implementation details for `RenderParagraph`, maybe we should push this logic down into TextPainter.  It would be sensible for TextPainter's layout function to work based on a BoxConstraints rather than having separate values for each of the fields on BoxConstraints.  Depending on where BoxConstraints is implemented, we might need to push that whole object down into the painting layer, but that's fine.

Hixie replied:
BoxConstraints extends Constraints which is defined in object.dart, but it doesn't currently reference RenderObject in any way, so I guess it could be pushed to painting/ safely. If you do this, put it in a new file, add it to painting,dart, and check the examples and tests to make sure anyone who uses it now imports painting.dart. (Do this in a separate PR if you do it.)

Hixie commented on sky/packages/sky/lib/src/fn3/editable_text.dart:
```diff
@@ -207,29 +195,67 @@ class EditableTextState extends State<EditableText> {
+  final SizeChangedCallback onContentSizeChanged;
+  final Offset scrollOffset;
+
+  RenderEditableParagraph createRenderObject() => new RenderEditableParagraph(
```
style nits:

- only use => if the expression fits on the same line. Otherwise, use a block with return.
- If you have a newline after a (, {, or [, you should have a newline before the corresponding ), }, or ]. (In fact in general, whitespace around open/close punctuation should be symmetric.)
- Indent two spaces from the previous indent.

Hixie commented on sky/packages/sky/lib/src/fn3/editable_text.dart:
```diff
@@ -207,29 +195,67 @@ class EditableTextState extends State<EditableText> {
-        [composingStyle, value.textInside(value.composing)],
-        value.textAfter(value.composing)
+      TextStyle composingStyle = style.merge(
+          const TextStyle(decoration: underline));
```
here's another example where you'd want to put the ) on a new line

Hixie commented on sky/packages/sky/lib/src/fn3/editable_text.dart:
```diff
@@ -207,29 +195,67 @@ class EditableTextState extends State<EditableText> {
+          new PlainTextSpan(value.textBefore(value.composing)),
+          new StyledTextSpan(composingStyle, [
+              new PlainTextSpan(value.textInside(value.composing))
+          ]),
```
this is good style

Hixie commented on sky/packages/sky/lib/src/fn3/input.dart:
```diff
@@ -18,14 +21,16 @@ typedef void StringValueChanged(String value);
     this.onChanged,
     this.keyboardType: KeyboardType.TEXT
-  }): super(key: key);
+  }): super(key: key,
```
put a newline after the (, then indent two spaces for the arguments, then ) on a line aligned with the }.

jason-simmons commented:
Addressed code review feedback - PTAL.

I made RenderParagraph._allowLineWrap final.  allowLineWrap was added so that RenderEditableParagraph could force text to render on a single line and then apply clipping.  I didn't envision it as a mutable property with a public API.

Hixie commented on sky/packages/sky/lib/src/fn3/editable_text.dart:
```diff
@@ -207,29 +196,72 @@ class EditableTextState extends State<EditableText> {
+        ]),
+        new PlainTextSpan(value.textAfter(value.composing))
       ]);
     } else {
```
No need for the else here, since the previous block returned.

Hixie commented on sky/packages/sky/lib/src/fn3/input.dart:
```diff
@@ -100,12 +112,17 @@ class InputState extends State<Input> {
+    );
+
     return new Listener(
       child: input,
```
I'd just inline container and input into the return statement here. Not sure why it was done separately before. It's easier to visualise the tree if it's all inline.

Hixie commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,66 @@ export 'package:sky/src/painting/text_painter.dart';
 
-  RenderParagraph(TextSpan text) : _textPainter = new TextPainter(text) {
+  RenderParagraph(
+      TextSpan text,
```
Indent two characters, not four. Put spaces around the named arguments.

(See the style guide for details.)

Hixie commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,66 @@ export 'package:sky/src/painting/text_painter.dart';
+  final TextPainter textPainter;
+
+  // Whether the text should be allowed to wrap to multiple lines.
+  final bool _allowLineWrap;
```
Why is this an immutable field? I think it's more idiomatic to use a getter.
That is, have this here:

```dart
   bool get allowLineWrap => true;
```

...and then override it in the subclass to return false.

Hixie commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,66 @@ export 'package:sky/src/painting/text_painter.dart';
+    layoutText(constraints);
     return constraints.constrainWidth(
-        _applyFloatingPointHack(_textPainter.minContentWidth));
+        applyFloatingPointHack(textPainter.minContentWidth));
```
Put this all on one line or put the closing ) on its own line.

Hixie commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,66 @@ export 'package:sky/src/painting/text_painter.dart';
+    layoutText(constraints);
     return constraints.constrainWidth(
-        _applyFloatingPointHack(_textPainter.maxContentWidth));
+        applyFloatingPointHack(textPainter.maxContentWidth));
```
ditto

Hixie commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -15,59 +15,66 @@ export 'package:sky/src/painting/text_painter.dart';
+    layoutText(constraints);
     return constraints.constrainHeight(
-        _applyFloatingPointHack(_textPainter.height));
+        applyFloatingPointHack(textPainter.height));
```
ditto

Hixie commented on sky/packages/sky/lib/src/rendering/paragraph.dart:
```diff
@@ -80,15 +87,15 @@ class RenderParagraph extends RenderBox {
   void performLayout() {
-    _layout(constraints);
+    layoutText(constraints);
     // _paragraphPainter.width always expands to fill, use maxContentWidth instead.
```
replace this comment with "We use textPainter.maxContentWidth here, rather that textPainter.width, because the latter is the width that it used to wrap the text, whereas the former is the actual width of the text."

Hixie commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,115 @@
+import 'package:sky/src/rendering/paragraph.dart';
+import 'package:sky/src/rendering/proxy_box.dart' show SizeChangedCallback;
+
+const _kCursorGap = 1.0;
```
put comments after these constants documenting the units, e.g.:

`const _kCursorGap = 1.0; // pixels`

Hixie commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,115 @@
+/// RenderParagraph but also renders a cursor and provides support for
+/// scrolling.
+class RenderEditableParagraph extends RenderParagraph {
+
```
If you have two newlines after the {, put two newlines before the } at the end of the class.

Hixie commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,115 @@
+
+  Offset get scrollOffset => _scrollOffset;
+  void set scrollOffset(Offset scrollOffset) {
+    _scrollOffset = scrollOffset;
```
check that it hasn't changed

Hixie commented on sky/packages/sky/lib/src/fn3/input.dart:
```diff
@@ -125,4 +142,26 @@ class InputState extends State<Input> {
+    scrollTo(scrollBehavior.updateExtents(
+      contentExtent: _contentWidth,
+      containerExtent: _containerWidth,
+      scrollOffset: _contentWidth));
```
newline before last ), per style guide

Hixie commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,115 @@
+
+  double getMinIntrinsicWidth(BoxConstraints constraints) {
+    return constraints.constrainWidth(
+      super.getMinIntrinsicWidth(constraints) + _kCursorGap + _kCursorWidth
```
This isn't accurate. For example, consider the case where the text looks like this:

```
Hello Nice Big
World|
```

...where the cursor is the vertical bar. The minimum intrinsic width in this case is unaffected by the cursor.

If you only support single-line text, such that there's only one intrinsic width, then just have a single function that you call from both getMinIntrinsicWidth and getMaxIntrinsicWidth (and assert that textPainter.minContentWidth and textPainter.maxContentWidth are the same, and document that this is only valid because we don't wrap).

Hixie commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,115 @@
+
+    if (_contentSize == null || _contentSize != newContentSize) {
+      _contentSize = newContentSize;
+      if (onContentSizeChanged != null) {
```
remove the { } around the block if it's a one-line block.

Hixie commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,115 @@
+
+    context.canvas.save();
+
+    if (_contentSize.width > size.width) {
```
remove braces around one line blocks

Hixie commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,115 @@
+
+    if (_showCursor) {
+      Rect cursorRect =  new Rect.fromLTWH(
+        textPainter.maxContentWidth + _kCursorGap,
```
shouldn't you be applying the floating point hack here?

Hixie commented:
Mostly just style nits.

jason-simmons commented:
Updated the CL

* Moved applyFloatingPointHack rounding to the (min|max)ContentWidth and height properties in TextPainter
* Skipped the (minContentWidth == maxContentWidth) check in EditableParagraph.getIntrinsicWidth.  minContentWidth reflects the width if the text was wrapped at every point even if the TextPainter's maxWidth is set to infinity

abarth commented on sky/packages/sky/lib/src/rendering/editable_paragraph.dart:
```diff
@@ -0,0 +1,124 @@
+  void paint(PaintingContext context, Offset offset) {
+    layoutText(constraints);
+
+    context.canvas.save();
```
We only need to call save/restore when we call clipRect.  We should move those calls inside the conditional too.

jason-simmons replied:
Done

abarth commented:
LGTM

Hixie commented:
LGTM
